feat: windows support

ALameLlama opened the pull request:
This is will WIP and I think this code can be cleaned up a lot, Adds keyboard and mouse support for windows.

#14

ALameLlama committed:
feat: windows size support

ALameLlama committed:
feat: keyboard support

ALameLlama committed:
feat: reverse con settings on destruct

ALameLlama committed:
feat: mouse support

ALameLlama committed:
feat: prevent sending ctrl/alt/shift keys

ALameLlama committed:
feat: better mouse support and Fn keys

ALameLlama committed:
test: size from windows

dantleech commented:
Great!

ALameLlama committed:
feat: move to WindowsConsole class and some cleanup

ALameLlama committed:
feat: some more cleanup

ALameLlama committed:
fix: sending key behaviour

ALameLlama commented:
I've cleaned some of the code I already had and fixed up the phpstan/php-cs-fixer errors, but I am still not entirely happy with this.

in my simple testing app:
```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

use PhpTui\Term\Actions;
use PhpTui\Term\ClearType;
use PhpTui\Term\Event\CharKeyEvent;
use PhpTui\Term\Terminal;

$terminal = Terminal::new();

$terminal->execute(Actions::cursorHide());
$terminal->execute(Actions::alternateScreenEnable());
// $terminal->execute(Actions::enableMouseCapture());
$terminal->enableRawMode();

echo 'press q to quit' . PHP_EOL;
while (true) {
    while ($event = $terminal->events()->next()) {
        if (empty($event)) {
            continue;
        }

        var_export($event);

        if ($event instanceof CharKeyEvent) {
            if ($event->char == 'q') {
                break 2;
            }
        }
    }
}

$terminal->disableRawMode();
// $terminal->execute(Actions::disableMouseCapture());
$terminal->execute(Actions::alternateScreenDisable());
$terminal->execute(Actions::cursorShow());
$terminal->execute(Actions::clear(ClearType::All));
```

It's instant and everything works. but using the example demo on the php-tui package itself it's slow and the mouse doesn't work.

I'll still need to investigate why it's so slow using the example demo and see why the mouse doesn't work. I believe I'll need to change this to use the virtual inputs, but when I was playing around with this before, it caused other issues, and the focus event stopped working altogether.

ALameLlama commented:
I also noticed a weird issue where things gauge page or any animiation starts for like a second and then pauses. and if I press something, it will go for a few more frames and then pause again, so I think this is why it feels so slow on the demo and it's an issue with the renderer.

dantleech commented:
Tbh it's quite.slow on Linux too when really it should be more or less the same speed as Rust/Ratatui for what it does, but couldn't identify the bottleneck. Maybe a different issue though.

ALameLlama commented:
With my testing I have vagrant running a pretty clean ubuntu box and then my main PC is windows, the ubuntu box is instant but there is noticeable lag on the windows side, I think there is something weird happening with the renderer since they're both really quick when I am just dumping out the event loop when actions are pressed.

dantleech commented on src/WindowsConsole.php:
```diff
@@ -0,0 +1,163 @@
+        return new self();
+    }
+
+    public function SetConsoleMode(int $mode): void
```
why CamelCase and not camelCase?

dantleech commented on src/WindowsConsole.php:
```diff
@@ -0,0 +1,163 @@
+    {
+        /**
+         * @phpstan-ignore-next-line */
+        $this->ffi->SetConsoleMode($this->handle, $this->originalSettings);
```
is there a particular reason to do this here vs. `RawMode::disable()` ?

ALameLlama replied:
you're right, I was originally doing this because I was testing things. I'll remove these deconstrctures

ALameLlama committed:
fix: mouse toggle and remove deconstructure

ALameLlama commented:
From a code standpoint, I am pretty happy with this. The little testing app I created earlier seems to work great. I am getting all the key inputs events, mouse inputs events, like movement, clicking, dragging, and scrolling, focus events, fn keys, and modifiers are getting sent.

I am using a weird 55% keyboard, so I'll try to find a 100% keyboard to double-check that I've covered all the basics. I may have missed things but they'll more than likely need to get added into this switch case that'll handle it https://github.com/php-tui/term/pull/15/files#diff-6195a21f4f231aa63422d3332f3ebdb343d6d1d2210436c7ebe0cb3a609faa3dR98-R124

I've still having issues with the php-tui demo app, I'll move this to ready once I've done some testing with a full keyboard and had more of a look at the demo app that's included with php-tui


https://github.com/user-attachments/assets/0af4fc18-f918-43ef-bdda-44f8bb8c0ca8

ALameLlama committed:
fix: camelCasing

ALameLlama committed:
fix: IO blocking

ALameLlama committed:
feat: move Win size provider to window console api

first commit of the new year 🎉

ALameLlama commented:
This is a bit of a New Year's Eve update, but I've moved from getting the size from cmd to the window console API and fixed an IO blocking issue. This has fixed some of the weird performance and mouse issues I was having, which is awesome!

Since I've moved it to use the window console API, I've needed to remove the test since it's no longer testable.

https://github.com/user-attachments/assets/b3c55d89-386d-48bd-a211-fe7768ff6a3

I still want to look into more performance optimizations I can do using xdebug. I might move the Windows console class into a "singletons" so it's not creating new variables In memory each loop now that it's being used in the size check. but I think that is now most of the issues fixed just want to see if I can now speed it up more.

ALameLlama committed:
feat: make WindowsConsole class a singleton

ALameLlama committed:
feat: add missing keys

ALameLlama commented:
I've fixed up everything I can from the php-tui/term side now, I've tested a 100% keyboard and have fixed up anything I've noticed missing, and performance-wise is seems great. it's now the rendering that is slow and the slowness happens here https://github.com/php-tui/php-tui/blob/main/src/Display/Display.php#L87-L94

I was looking at traces from xdebug and noticed that mb_strwidth was much slower on windows and this seems to be a thing.
https://stackoverflow.com/questions/45028018/php-7-mb-multibyte-functions-are-60-slower-than-in-5-3-windows-only-issue

This was testing it on Windows vs my Ubuntu VM.
```
PS D:\dev-stuff> php .\testing-term-full\src\mb_testing.php
8.3.9

with encoding parameter:
- mb_strlen: 1,666 ms, result: 5
- mb_substr: 1,799 ms, result: う

with internal encoding:
- mb_strlen: 1,612 ms, result: 5
- mb_substr: 1,687 ms, result: う
```

```
 🕒 04:07 PM testing-term-full ❯ php src/mb_testing.php                                                                  vagrant in 🌐 vagrant in  master[!?] via 🐘 v8.3.15
8.3.15

with encoding parameter:
- mb_strlen: 3 ms, result: 5
- mb_substr: 8 ms, result: う

with internal encoding:
- mb_strlen: 3 ms, result: 5
- mb_substr: 7 ms, result: う
```

ALameLlama commented:
Let me know if you want me to make any changes :)

ALameLlama commented:
Also the resize event doesn't work https://github.com/php-tui/term/blob/main/src/EventProvider/SignalEventProvider.php#L26-L33

If get this to work, I'll need to change the Size from provider to keep trace of current and old size and then compare them each loop.

ALameLlama commented:
I didn't fully have xdebug disabled and updated my PHP  from 8.3.9 to 8.3.15 and the performance is a lot closer now.

```
PS D:\dev-stuff> php .\testing-term-full\src\mb_testing.php
8.3.15

with encoding parameter:
- mb_strlen: 5 ms, result: 5
- mb_substr: 10 ms, result: う
- mb_strwidth: 15 ms, result: 10

with internal encoding:
- mb_strlen: 4 ms, result: 5
- mb_substr: 10 ms, result: う
- mb_strwidth: 18 ms, result: 10
```

https://github.com/user-attachments/assets/14925922-e85f-4559-9b1a-3255890e5dfb

ALameLlama committed:
feat: better OS check

GianfriAur commented:
Hi @ALameLlama and @dantleech — just a friendly ping to check where this PR stands.

I'm the author of [`php-opcua/opcua-cli`](https://github.com/php-opcua/opcua-cli), a CLI tool for OPC UA industrial servers (PLCs, SCADA). I recently shipped an `explore` command built on `php-tui` that gives users an interactive TUI browser of the server address space — it works great on Linux and macOS, but for now I had to emit a clear "not yet supported on Windows" error on that platform. My audience is industrial automation engineers and a good chunk of them are on Windows workstations, so Windows support in `php-tui` is genuinely a feature that would unlock a real-world use case on my side.

Is there anything blocking the merge at this point, or is it waiting on review bandwidth? If it would help, I'm happy to give it a test drive on Windows 10 / 11 / Server and report back — not offering to take over or rush anything, just mentioning it in case an extra pair of hands on testing is useful.

Either way, thanks a lot for the work on this PR, and to @dantleech for maintaining `php-tui` in general. No pressure — happy to wait until it lands naturally.

ALameLlama commented:
Hey @GianfriAur It's been a while since I've worked on this but I still have a windows machine around to test this. From what I remember this was basically all working outside of the resize event due to it using a unix only method. There was some slowness (about 5x slower but I don't know if this is resolvable) and I traced it down to here https://github.com/php-tui/php-tui/blob/main/src/Display/Display.php#L87-L94

I'm more then happy to continue making any changes that are required. I think I was mostly seeking guidance from @dantleech about how we'd want to handle having a unix and windows variant of things like the stream reader, size provider etc and if they were happy with what I've provided here.

dantleech commented:
I think my main concern was about FFI not being enabled by default.. although from what I can see it's `preload` by default which [seems to imply](https://www.php.net/manual/en/ffi.configuration.php) that its available on the CLI?

ALameLlama commented:
Looks like it's enabled by default on CLI but not on FPM which is perfectly fine for us, since this is getting traction, I'll have another look over the PR to see if I'm still happy with it or if I can come up with better ways of handling things now

GianfriAur commented:
Thank you both for the really quick responses - didn't expect to hear back the same day, genuinely appreciated.

No rush on my side - just happy to know the PR isn't stalled and that you're willing to take another look when you have the time.  I remain available if you need help with development or testing.

Thanks again

dantleech commented:
Testing would be appreciated, I don't have access to a Windows machine and am not sure how to go about setting up or debugging a windows build in CI. So any early feedback would be good :)

GianfriAur commented:
> Testing would be appreciated, I don't have access to a Windows machine and am not sure how to go about setting up or debugging a windows build in CI. So any early feedback would be good :)

  Before moving on with more testing, I put together #17  - a CI PR that sets up automated tests across Linux, macOS and Windows for all currently supported PHP versions (8.1 -> 8.5). Once that's merged, we'll have proper cross-platform coverage to validate the Windows work here.

ALameLlama committed:
fix: size for windows

ALameLlama commented:
Hey, I noticed I was using the "A" (ANSI) variant of the FFI functions instead of "W" (Unicode) and I could handle repeating character better. I'll fixes these up and keep testing

ALameLlama committed:
refactor: windows to handle unicode and better performance

ALameLlama commented:
I've cleaned up the code so we're not passing CData around as much so it should be easier to follow along with what is happening, I've added PHPStan types too.

Before it was processing 1 input at a time but windows includes a repeating amount so this should be a little faster now when you're hold down a character or mouse inputs.

and it should now be able to handle more then just anscii too.

https://github.com/user-attachments/assets/6ec8ba13-389e-43ee-9648-d4230949069c

I'm still looking into if resize events are possible and I'll have another look on the rendering side in php-tui to see if we have make that any faster too

ALameLlama committed:
feat: windows resize event

dantleech commented on src/Painter/AnsiPainter.php:
```diff
@@ -79,26 +81,41 @@ private function drawCommand(Action $action): void
-            ])));
-
+            if (Terminal::isWindows()) {
+                $windowsConsole = WindowsConsole::getInstance();
```
maybe allow this to be injected?

```
public function __construct(Writer $writer, ?WindowsConsole $windowsConsole = new WindowsConsole()) {
}
```
(or equivalent for the supporter PHP version)

ALameLlama replied:
I can do something like
```
    public function __construct(private readonly Writer $writer, private readonly ?WindowsConsole $windowsConsole)
    {
    }

    public static function new(Writer $writer, ?WindowsConsole $windowsConsole = null): self
    {
        return new self($writer, $windowsConsole);
    }
```
    
but then I'd need to update the Terminal.php to do.
```
            $painter ?? AnsiPainter::new(StreamWriter::stdout(),  self::isWindows() ? WindowsConsole::new() : null),
```

Since WindowsConsole requires FFI and `$ffi = FFI::cdef($header, 'kernel32.dll');` I don't want to cause any headaches when it's on non Windows systems

dantleech replied:
I see, I guess a singleton is fine then.

ALameLlama committed:
fix: tab modifiers for windows

ALameLlama commented:
@dantleech I haven't included it in this branch but are you happy with this change? https://github.com/php-tui/term/commit/34dfdf194e7f3d6d46a97e3a76615ffd5eb367cb

I narrowed down the slowness I was seeing to here, this array can contain 1000+ actions and each of these is triggering a draw.

Doing them in a buffer like this greatly improved the FPS for me.

There is another painter that should also need to get a similar update, we could also do it in chunks so instead of doing a write for every item we should instead do it for every x amount instead of waiting for all of them to get into the buffer.

e.g
```php
    public function paint(array $actions): void
    {
        $buffer = '';

        foreach ($actions as $action) {
            $buffer .= $this->drawCommand($action);

            if (strlen($buffer) >= 8192) {
                $this->writer->write($buffer);
                $buffer = '';
            }
        }

        if ($buffer !== '') {
            $this->writer->write($buffer);
        }
    }
```

https://github.com/user-attachments/assets/25171950-cbfc-42a7-8973-35b097fcba64

ALameLlama committed:
feat: fix phpstan and php-cs-fixer

ALameLlama commented:
Outside of any feedback I receive this PR, I think most of the issue have been resolved, It would be great if @GianfriAur could have a play around with it and let me know they run into any issues.

The only other thing is what I mentioned in https://github.com/php-tui/term/pull/15#issuecomment-4319665293 with the buffer writes to improve performance

dantleech commented:
> I narrowed down the slowness I was seeing to here, this array can contain 1000+ actions and each of these is triggering a draw.

Yes -- maybe as a separate PR as I guess it's adjacent to Windows support?

ALameLlama commented:
> > I narrowed down the slowness I was seeing to here, this array can contain 1000+ actions and each of these is triggering a draw.
> 
> Yes -- maybe as a separate PR as I guess it's adjacent to Windows support?

I did branch it off and I'll create another PR once we're happy with this one, Next time I have my linux machine I'll test it out since I think it should also help with performance on linux too since it should just be less draw calls

GianfriAur commented:
> Outside of any feedback I receive this PR, I think most of the issue have been resolved, It would be great if @GianfriAur could have a play around with it and let me know they run into any issues.
> 
> The only other thing is what I mentioned in https://github.com/php-tui/term/pull/15#issuecomment-4319665293 with the buffer writes to improve performance 

Of course, very gladly. Thank you.

GianfriAur commented:
> Outside of any feedback I receive this PR, I think most of the issue have been resolved, It would be great if @GianfriAur could have a play around with it and let me know they run into any issues.
> 
> The only other thing is what I mentioned in [#15 (comment)](https://github.com/php-tui/term/pull/15#issuecomment-4319665293) with the buffer writes to improve performance

Hi everyone, as promised I did some testing on Windows, 

First important note: before starting any instance of the FFI class on Windows, you should verify that the extension (ffi) is active, and report it via a managed error, perhaps directly at startup? Otherwise, the user would see the error stack reporting that the FFI class simply does not exist.

Then I add that I am not very experienced with php on windows, I try to avoid it as much as possible, but once I configured everything, I could not get the demo in `php-tui/php-tui` to run receiving this error:

``` php .\example\demo\bin\demo
PHP Fatal error:  Uncaught FFI\ParserException: Undefined C type "CONSOLE_SCREEN_BUFFER_INFO" at line 1 in C:\Users\.......\vendor\php-tui\term\src\WindowsConsole.php:59
 ``` 
same error with both cmd and Powershell
PHP 8.5.5

Maybe I did something wrong? Or is there some special configuration I need to do? @ALameLlama

ALameLlama commented:
@GianfriAur For my testing I was using Windows 11 with PHP 8.5.5 installed via choco I think the only change I needed to do was enable some extensions like intl and mb I didn't really need to do any extra configuration.

Was your testing also done on Windows 11? if not I can spin up a VM and try and test is on some other versions

Edit: I was dumb and broke this last commit when I was trying to make PHPStan happy can you try it again? I've also added a throw when the extension isn't loaded.

Unfortunately I don't this PHP lets you do per platform requirements in composer where normally this would get caught

ALameLlama committed:
fix: ffi struct and added extension check

GianfriAur commented:
> @GianfriAur For my testing I was using Windows 11 with PHP 8.5.5 installed via choco I think the only change I needed to do was enable some extensions like intl and mb I didn't really need to do any extra configuration.
> 
> Was your testing also done on Windows 11? if not I can spin up a VM and try and test is on some other versions

@ALameLlama I used a physical PC with W11 and installed php from the official command line.

The change you made worked like magic. Now I confirm that everything works, both cmd and powershell.

the only thing that should be put in `php-tui/php-tui` where the `IntlChar` class is used like (BrailleGrid) is the control if the `intl` extension is activated @dantleech

dantleech commented:
@GianfriAur I've updated composer to require the INTL extension for now in addition to updating PHPStan to 2.0: https://github.com/php-tui/php-tui/pull/235

GianfriAur commented:
> @GianfriAur I've updated composer to require the INTL extension for now in addition to updating PHPStan to 2.0: [php-tui/php-tui#235](https://github.com/php-tui/php-tui/pull/235)

@dantleech if it's okay with you, as soon as I have a second I can start fixing phpstan and do a pr
